### PR TITLE
IE8 compatibility

### PR DESCRIPTION
**Pull request by salzhrani**
First batch of changes ... feedback is greatly appreciated

### CODE REVIEW

**Comment by oneeman** [comment]
Awesome, thanks for tackling this!  It'd be great to have IE8 passing.  Here's what I've got:

* For skipping tests involving svg, if possible place those tests all together so they can be skipped with just one such conditional per file.  And use the same test condition for consistency (unless there's a reason not to which I'm not aware of).  What I mean is in some you check whether `createElementNS` exists and in others whether a `div` has `namespaceURI`, instead use just one of those two if possible.  Perhaps even move that feature detection to the test helper and give its result a descriptive name, then use that boolean in the files with tests.
* Also for the skipping, I noticed the style used so far has been to not indent the contents of the `if` for skipping at all.  I assume we want to keep it that way.
* For the fixes for `textContent`,  don't use the ternary operator or `if`'s  since it just makes less clear than it needs to be.  Instead, if the one that works for IE8 works for other browsers, just use that.

Other comments will be as line comments.

File: packages/htmlbars-syntax/lib/utils.js [hunk 1]
```diff
@@ -17,7 +17,14 @@ export function parseComponentBlockParams(element, program) {
...
 
-  var asIndex = attrNames.indexOf('as');
+  var asIndex = -1;
+
+  for (i = 0; i < attrNames.length; i++) {
+    if (attrNames[i] === 'as') {
+      asIndex = i;
+      break;
+    }
+  }
```
Comment (oneeman) [inline packages/htmlbars-syntax/lib/utils.js]: Wow, I had no idea.  For the `for` loop, please use `i < l` rather than `i < attrNames.length`.  It's customary to do this for performance reasons, and you often see stuff like `for (var i=0, l=items.length; i < l; i++)` in the ember codebase.  In this case, we already have `l` from above.

File: packages/htmlbars-compiler/tests/html-compiler-test.js [hunk 2]
```diff
@@ -199,12 +212,22 @@ function shouldBeVoid(tagName) {
...
+  var closing;
+  if (div.innerHTML.indexOf(':') >= 0) {
+    tag = '<:' + tagName + ' data-foo="bar">';
+    closing = '</:' + tagName + '>';
+  } else {
+    tag = '<' + tagName + ' data-foo="bar">';
+    closing = '</' + tagName + '>';
+  }
+    
   var extra = "<p>hello</p>";
```
Comment (oneeman) [inline packages/htmlbars-compiler/tests/html-compiler-test.js]: Please implement the fix for this problem in `normalizeInnerHTML` in the `ie8InnerHTML` clause.  Should be doable with a regex search like the other fixes there.

**Comment by oneeman** [comment]
Also, could you please remove the `allow_failures` section in `.travis.yml`?  IE8 is the last one failing, so we shouldn't need that section anymore after this PR.

File: packages/htmlbars-syntax/lib/utils.js (hunk 1 above)
Comment (mixonic) [inline packages/htmlbars-syntax/lib/utils.js]: @salzhrani I encourage you to create an `indexOfArray` method that is the native version if available, and a poly-fill if not.

I suggest reviewing [ember's implementation](https://github.com/emberjs/ember.js/blob/d677cbb0c51eee34910ab4075bc53c938d0e3055/packages/ember-metal/lib/array.js#L62) for inspiration!

File: packages/morph/lib/dom-helper.js [hunk 3]
```diff
@@ -141,7 +141,7 @@ prototype.childAt = function(element, indices) {
 };
 
 prototype.childAtIndex = function(element, index) {
-  return element.childNodes.item(index);
+  return element.childNodes.length > index ? element.childNodes.item(index) : null;
```
Comment (mixonic) [inline packages/morph/lib/dom-helper.js]: Why is this needed?

File: packages/morph/lib/dom-helper.js [hunk 4]
```diff
@@ -153,7 +153,11 @@ prototype.setAttribute = function(element, name, value) {
 };
 
 prototype.setAttributeNS = function(element, namespace, name, value) {
-  element.setAttributeNS(namespace, name, String(value));
+  if (element.setAttributeNS) {
```
Comment (mixonic) [inline packages/morph/lib/dom-helper.js]: Is this feature detection need on an element-by-element basis, or can we treat it as a browser feature detect?

```js
// I'm sure this detect can be written better, see the top of the file for others
if (document && document.createElement('div').setAttributeNS)  {
  prototype.setAttributeNS = function(element, namespace, name, value) {
    element.setAttributeNS(namespace, name, String(value));
  }
} else {
  prototype.setAttributeNS = function(element, namespace, name, value) {
    element.setAttribute(name, String(value));
  }
}
```

File: packages/htmlbars-compiler/tests/fragment-test.js [hunk 5]
```diff
@@ -178,13 +183,13 @@ test('test auto insertion of text nodes for needed edges a fragment with morph m
...
   equal(t.nodeType, 3);
-  equal(t.textContent , '');
+  equal(t.textContent !== undefined ? t.textContent : t.toString() , '');
   equal(morphs[1].start, null);
   equal(morphs[1].end, null);
 
   equal(morphs[2].start, morphs[1].parent());
   equal(morphs[2].end.nodeType, 3);
-  equal(morphs[2].end.textContent, '');
+  equal(morphs[2].end.textContent  !== undefined ? morphs[2].end.textContent : morphs[2].end.toString(), '');
```
Comment (mixonic) [inline packages/htmlbars-compiler/tests/fragment-test.js]: These assertions with ternary statements are really confusing. They should be the same for all browsers.

**Comment by mixonic** [comment]
Please do group the SVG tests together. I generally prefer to not indent in the test files because it keeps diffs much easier to read.

**Comment by mixonic** [comment]
@salzhrani if you have time this week, we will need to get IE8 fixes in before the weekend. It would be a huge help if we could just merge this PR!

**Comment by salzhrani** [comment]
I understand and I am trying my best. I have time on Friday and I believe I can get it done by then.

**Comment by oneeman** [comment]
Much appreciated!  Since I've got time this week, how about I work on the remaining failures (which don't look too bad... at least they're failing rather than dying like the ones you debugged were!), and when you have time you can finalize what you already have here?  Then, between the two, IE8 should be green.

**Comment by salzhrani** [comment]
I'll give be it a swing tonight to incorporate the feedback and fix a few more issues and @oneeman can take it from there.

**Comment by oneeman** [comment]
Awesome.  Please rebase and squash as well as there have been a lot of commits to master the last few days.  I made a bit of progress last night, which you can check out at: https://github.com/oneeman/htmlbars/commits/more-ie8, so as to avoid duplicating effort when you work on remaining issues.  Once your changes are done I can rebase mine as well and keep going.  And I'll be on IRC tonight (CST).

File: packages/morph/lib/dom-helper.js (hunk 3 above)
Comment (oneeman) [inline packages/morph/lib/dom-helper.js]: This should take care of https://github.com/tildeio/htmlbars/issues/259

**Comment by salzhrani** [comment]
rebased and squashed. Failed tests are down to 27, Thank you all for the feedback. @oneeman appreciate you tackling the remaining issues.

File: packages/morph/tests/dom-helper-test.js [hunk 6]
```diff
@@ -56,9 +57,9 @@ test('#setAttribute', function(){
 
   node = dom.createElement('input');
   dom.setAttribute(node, 'disabled', true);
-  equalHTML(node, '<input disabled="true">');
+  equalHTML(node, ie8 ? '<input disabled>' : '<input disabled="true">');
```
Comment (mixonic) [inline packages/morph/tests/dom-helper-test.js]: This can just assert `ok(!!node.getAttribute('disabled'))`.

Then we can avoid the export of `ie8` (we don't detect browsers anywhere else, just features) and use of it only in these tests.

**Comment by salzhrani** [comment]
This is as far as I can take this PR given the required expertise and time constraints. Thank you all

**Comment by oneeman** [comment]
Absolutely, thank you!  I noticed you managed to do some additional clean-up and even fix some more failing tests.  We can merge and I'll continue on top of this in another PR.

**Comment by mmun** [comment]
Great work! :+1:

**Comment by mixonic** [comment]
@salzhrani thanks for helping to shoulder the burden of IE8 with us :-) I appreciate your patience with my comments.

**Comment by salzhrani** [comment]
It is a pleasure. I look forward to more comments .
